### ISSUE TITLE
YOLO26m vs YOLO11m: Lower Recall-Confidence Curve Despite Identical mAP — Confidence Calibration Gap?

### ISSUE THREAD

**Issue by marcosalemany** [ultralytics/ultralytics#24062]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

## Environment

```
Ultralytics 8.4.14 🚀 Python-3.12.3 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 5090, 32075MiB)
Setup complete ✅ (16 CPUs, 31.0 GB RAM, 935.3/936.8 GB disk)
OS                     Linux-6.8.0-100-generic-x86_64-with-glibc2.39
Environment            Linux
Python                 3.12.3
Install                pip
RAM                    30.99 GB
CPU                    Intel Xeon w3-2435
CPU count              16
GPU                    NVIDIA GeForce RTX 5090, 32075MiB
GPU count              1
CUDA                   12.8
numpy                  ✅ 1.26.4
torch                  ✅ 2.8.0
torchvision            ✅ 0.23.0
```

---

## Task & Models

- **Task:** Object detection (multi-class, 5 categories)
- **Models:** YOLO11m vs YOLO26m
- **Dataset:** Custom dataset, COCO format
- **Input size:** 640×640
- **Deployment:** DeepStream 8.0 + TensorRT FP16, multi-camera setup, dense scenes

---

## Training Setup

Both models trained on the **same dataset** with **identical hyperparameters**:

```yaml
epochs: 300  (early stopping, YOLO11 stopped at 271, YOLO26 at 246)
batch: 16
imgsz: 640
optimizer: AdamW
lr0: 0.001
lrf: 0.01
```

No differences in augmentation or data splits.

---

### Metrics summary (best epoch, val set)

| Metric | YOLO11m | YOLO26m | Delta |
|---|---|---|---|
| Best mAP50-95 | 0.5723 | 0.5735 | **≈ identical** (+0.0012) |
| Best mAP50 | 0.7466 | 0.7440 | –0.003 (negligible) |
| Best Recall (val @ conf=0.25) | 0.7234 | 0.7076 | **–1.6 pp** |
| Val box loss (final epoch) | 1.271 | 1.485 | **+17%** |
| Val cls loss (final epoch) | 1.123 | 1.476 | **+32%** |
| Best mAP epoch | 221 | 196 | YOLO26 converges ~25 ep earlier |

### Recall-Confidence curves

The curves start nearly identical at conf=0.000 (0.81 vs 0.82), but YOLO26's curve degrades noticeably faster as the threshold increases. At conf=0.25 the gap is approximately 4–6 pp on all classes, and larger on the primary class.

*(Recall-Confidence plots for both models attached below)*

---

## Hypothesis: Confidence Calibration Gap, Not Ranking Gap

The apparent contradiction — same mAP but worse Recall-Confidence — points to a **calibration issue**, not a ranking issue:

**mAP is blind to absolute confidence values.** It ranks all predictions by score and measures the area under the P-R curve. Two models with identical prediction ordering but different absolute scores will produce identical AP. This explains why mAP50-95 is essentially the same.

**The Recall-Confidence curve is sensitive to absolute values.** It answers "how many objects survive threshold T?" — which depends directly on the absolute scores. If YOLO26 systematically outputs lower/compressed confidence scores, the curve drops faster even if the detector is equally capable in terms of ranking.

**The val losses are the clearest signal.** YOLO26 has persistently higher `val/box_loss` (+17%) and `val/cls_loss` (+32%) across all ~250 epochs, despite converging to the same mAP plateau. Since loss penalizes absolute prediction error (not just ranking), this persistent gap is consistent with a model that produces systematically less-calibrated confidence scores.

**Suspected structural cause:** with the one-to-one head, there is a single prediction slot per object. In one-to-many (YOLO11 + NMS), several anchors compete and the highest-confidence survivor sets the output score — effectively the *maximum* across redundant predictions. With one-to-one, if the single slot carries uncertainty, the output score reflects that uncertainty directly with no redundancy to boost it. The one-to-many auxiliary head during training is meant to compensate for this via richer gradient supervision, but the val loss gap suggests it may not fully close the calibration difference at inference.

---

## Deployment Context and Parser Behavior

Our inference stack uses [[marcoslucianops/DeepStream-Yolo](https://github.com/marcoslucianops/DeepStream-Yolo)](https://github.com/marcoslucianops/DeepStream-Yolo) with a custom `NvDsInferParseYolo` parser. Both models use identical parser code and FP16 TensorRT engines.

**YOLO26m nvinfer config — `cluster-mode=4` (NMS-free, appropriate for one-to-one head)**
```ini
[property]
gpu-id=0
net-scale-factor=0.0039215697906911373
model-color-format=0
onnx-file=yolo26m.onnx
model-engine-file=yolo26m_b8.engine
labelfile-path=labels.txt
batch-size=8
network-mode=2
num-detected-classes=5
interval=0
gie-unique-id=1
process-mode=1
network-type=0
cluster-mode=4
maintain-aspect-ratio=1
symmetric-padding=1
parse-bbox-func-name=NvDsInferParseYolo
custom-lib-path=libnvdsinfer_custom_impl_Yolo.so
engine-create-func-name=NvDsInferYoloCudaEngineGet

[class-attrs-all]
pre-cluster-threshold=0.3
topk=10000
```

**YOLO11m nvinfer config — `cluster-mode=2` (standard DeepStream NMS)**
```ini
[property]
gpu-id=0
net-scale-factor=0.0039215697906911373
model-color-format=0
onnx-file=yolo11m.onnx
model-engine-file=yolo11m_b8.engine
labelfile-path=labels.txt
batch-size=8
network-mode=2
num-detected-classes=5
interval=0
gie-unique-id=1
process-mode=1
network-type=0
cluster-mode=2
maintain-aspect-ratio=1
symmetric-padding=1
parse-bbox-func-name=NvDsInferParseYolo
custom-lib-path=libnvdsinfer_custom_impl_Yolo.so
engine-create-func-name=NvDsInferYoloCudaEngineGet

[class-attrs-all]
pre-cluster-threshold=0.3
nms-iou-threshold=0.45
topk=10000
```

Both models are configured correctly for their respective architectures. The recall gap in the Recall-Confidence curves is therefore not attributable to any NMS asymmetry between configurations — it appears to reflect a genuine difference in confidence score calibration from training.

One deployment-specific observation worth noting: in dense scenes with high object counts, we found that naively lowering `pre-cluster-threshold` to recover recall actually worsens detection. The reason is parser buffer saturation — the `NvDsInferParseYolo` parser reads the output tensor sequentially (raster-scan order, top→bottom), fills its internal buffer with low-confidence noise from the upper image region, and drops real detections in the lower region before NMS is reached. This makes the calibration gap particularly impactful in production: a model with compressed confidence scores hits that saturation threshold sooner and at a higher operational threshold than a well-calibrated model. YOLO26's one-to-one head (no duplicate proposals) is architecturally advantageous here, as it eliminates the buffer saturation problem entirely — but only if the confidence calibration gap is addressed.

---

## Questions

1. **Is the confidence calibration gap an expected consequence of the one-to-one head design**, even with the one-to-many auxiliary head providing supervision during training? Is the +17%/+32% val loss gap something the team has observed across datasets?

2. **Are there recommended training-time approaches to improve calibration** within the NMS-free setup — e.g. adjusting `cls_pw`, modified loss weights, label smoothing, or temperature scaling post-hoc?

3. **Is the persistent val loss gap structural** (inherent to the one-to-one architecture), or could it indicate a domain-specific tuning issue that could be addressed with hyperparameter changes?

---

## Attachments

- `results_yolo11m.csv` — full training metrics, 271 epochs
- `results_yolo26m.csv` — full training metrics, 246 epochs
- Recall-Confidence curve plots (both models, same val set)

*(CSV files and plots attached to this issue)*
### YOLO11M
<img width="2250" height="1500" alt="Image" src="https://github.com/user-attachments/assets/db9f3082-0fb6-4e12-aed7-3f386f71295b" />

[yolo11_results.csv](https://github.com/user-attachments/files/26347892/yolo11_results.csv)

### YOLO26M

<img width="2250" height="1500" alt="Image" src="https://github.com/user-attachments/assets/e5cb7750-a8a4-444d-a1f6-f19b066c8d4e" />

[yolo26_results.csv](https://github.com/user-attachments/files/26347931/yolo26_results.csv)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @marcosalemany, thank you for your interest in Ultralytics 🚀 and for the very detailed report. This is an automated response to help triage your issue, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a custom training and evaluation ❓ question, please provide as much information as possible, including dataset image examples, exact training commands, validation commands, export/inference commands, and any relevant logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
If you believe there may also be a 🐛 bug involved in the training, validation, export, or deployment pipeline, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) showing the smallest possible setup that reproduces the behavior consistently.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify the issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help the team review this faster 🔍, if possible please also share:
- The exact `yolo` train/val/export commands used for both YOLO11m and YOLO26m
- Whether the same data YAML, augmentations, seed, and preprocessing pipeline were used end-to-end
- Whether the same confidence and IoU settings were used during validation and deployment comparisons
- A minimal reproducible subset or script if the confidence/recall behavior can be reproduced consistently outside the full production pipeline

**Comment by Y-T-G** [comment]
The losses are not comparable because they are different architectures with different loss functions.

I think the recall drop is expected with NMS-free head. You can try training for longer to improve recall.

**Comment by marcosalemany** [comment]
Thanks for the clarification 🙏  — understood on the loss functions. You're right that **comparing val/box_loss** directly across architectures with different loss formulations **is misleading**.

**On training longer:** the model stopped at epoch 246 via early stopping (patience=50) because mAP50-95 had plateaued. The concern is that **patience is tied exclusively to mAP** — if confidence calibration continues improving after the mAP plateau (which seems plausible given the one-to-one head needs more signal to learn well-calibrated unique assignments), early stopping would cut it regardless.

I am **considering two approaches to decouple patience from mAP** and allow calibration to keep improving:

- Option 1 — composite fitness including recall at a fixed threshold

- Option 2 — ECE (Expected Calibration Error)-penalized fitness to directly optimize calibration

In parallel, we are also running additional training experiments with different hyperparameter configurations (augmentation strategy, mosaic, copy_paste, freeze, loss weights) to evaluate whether the calibration gap can be reduced from the training side independently of epoch count. We'll report back with results.

**Before implementing either approach, two questions for the team:**

1. Is there a recommended way to expose calibration-aware early stopping in Ultralytics, or is the callback approach above the intended path?
2. For YOLO26m specifically, is there a known minimum epoch count or patience recommendation for dense-scene datasets where the one-to-one head needs more time to converge on calibration?

**Comment by burak43** [comment]
Experiencing the very same problem. My findings are in alingment with OP's hypothesis. YOLO26 is performing worse on distant-objects due to confidence jittering which do not only occur on distant-objects. Please share your advice if you have any.

**Comment by glenn-jocher** [comment]
Thanks, this is useful context — today the right path for calibration-aware stopping is a custom callback or trainer change, and we do not have a fixed `epochs` or `patience` target for dense-scene YOLO26 because it is dataset-dependent; for this case I’d compare validation with `end2end=True` and `end2end=False` as shown in the [end-to-end detection guide](https://docs.ultralytics.com/guides/end2end-detection/) to separate one-to-one score calibration from core detector quality, and if others are seeing the same distant-object effect, please attach a small reproducible sample plus size-binned metrics so we can verify whether it is dataset-specific or broader. ([docs.ultralytics.com](https://docs.ultralytics.com/guides/end2end-detection/))

**Comment by burak43** [comment]
@glenn-jocher I've had mesured the KPIs with `e2e=True` (i.e. `o2o` head) and `e2e=False` (i.e. `o2m` head) to see if this is a core detector quality issue or not as you mentioned. `o2m` head is _slightly_ better at detecting but even that one fails to detect distant-objects (or detects with very low confidences) **compared to YOLOv11** and confidence instability/jittering persists on both. I believe that it may be due to ProgLoss implementation that might leave both heads/backbone as not fully converged or it can be something else idk. I'd be glad for your suggestions.

**Comment by Y-T-G** [comment]
@burak43 Are you using pretrained YOLO26 models? Which size of the model are you using?

**Comment by burak43** [comment]
@Y-T-G I'm using a custom model based on the nano variant of pose architecture but my observation are based on the detection performance only.
